### PR TITLE
Mouseflow

### PR DESCRIPTION
**Pull request by lc512k**
@adambraimbridge 

- [x] translating the query the user entered in the box into a `keenIO.Query('extraction`
- [x] Extracting an array of spoorIds from that, to POST to the next-perspective endpoint
- [x] Building a MVP table of `spoorId/URL` | `length` | `something` from those results
- [x] Exposing next-perspective endpoint (separate repo)
- [x] Make it look nice
- [x] n-perspective
- [x] Sortable table

Pending
https://github.com/Financial-Times/next-beacon-v2/issues/277

### CODE REVIEW

File: client/components/chart-ui.js [hunk 1]
```diff
@@ -30,17 +30,14 @@ function copyData (ev) {
...
+function getChartContainer () {
+	return document.querySelector('.chart');
 }
 
 function reprint (ev) {
 	ev.preventDefault();
-	const container = getChartContainer(ev.target);
-	const configure = getConfigurator(container.querySelector('.chart__configurator'), ev.target.name);
+	const container = getChartContainer();
+	const configure = getConfigurator(document.querySelector('.chart__configurator'), ev.target.name);
```
Comment (lc512k) [inline client/components/chart-ui.js]: @adambraimbridge: the markup changed a bit and configurator is now outside of chart (container)

File: client/components/range.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	const output = document.querySelector('.extract__numberOfEvents');
+
+	if (slider && output) {
+		output.value = slider.value;
+	}
+}
+
+export function init () {
+	const delegate = new Delegate(document.body);
+	delegate.on('change', '.extract__numberOfEvents-range', handleRangeChange);
```
Comment (adambraimbridge) [inline client/components/range.js]: Note: Review this event

File: client/components/recordings.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+function startLoading (el) {
+	el.classList.add('chart--loading');
+}
+
+function doneLoading (el) {
+	el.classList.remove('chart--loading');
+}
+
+function predictSluggishness (period, latest, messagesEl) {
```
Comment (adambraimbridge) [inline client/components/recordings.js]: idea: "sample rate" variable. e.g. 1% means cast a wider net

File: client/components/recordings.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	else if (period <= 14 & latest <= 1000) {
+		message = 'This query will take a few minutes to complete ⌛ (Try a smaller timeframe for faster results)';
+	}
+
+	messagesEl.innerHTML = message;
+}
+
+export function getRecordings ({el, queryStr, eventLimit, messagesEl, userTimeframe, configuration, chartName}={}) {
+	// TODO handle concatenated queries
```
Comment (adambraimbridge) [inline client/components/recordings.js]: note: check out the other aggregators https://github.com/Financial-Times/keen-query#aggregators

File: client/components/recordings.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	if (configuration && chartName) {
+		kq = configuration(retrieveKq(chartName));
+	}
+	else {
+		kq = KeenQuery.build(queryStr);
+	}
+
+	const query = new keenIO.Query('extraction', {
+		timeframe: kq.timeframe,
+		event_collection: kq.query.event_collection,
```
Comment (adambraimbridge) [inline client/components/recordings.js]: see note above re: aggregtors

File: client/components/recordings.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+		// FROM THE MOUSEFLOW DOCS
+		// fromdate: The start date of the query. The actual start time is midnight on the selected date, according to the user’s selected time zone.
+		const fromtime = new Date(now - period * day - day);
+		const fromdate = formatDate(fromtime);
+
+		let entry;
+
+		for (const filter of kq.filters) {
+			if (filter.property_name.toLowerCase().indexOf('normalizedpath') !== -1) { //TODO other paths
```
Comment (adambraimbridge) [inline client/components/recordings.js]: bump

File: client/components/recordings.js [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+				const tableEl = window.$('.recordings');
+				tableEl.stupidtable();
+
+				return true;
+			})
+			.catch((error) => {
+				doneLoading(el);
+				el.innerHTML = error;
+				return;
```
Comment (adambraimbridge) [inline client/components/recordings.js]: todo: make the error message pretty

File: views/extract.html [hunk 8]
```diff
@@ -26,10 +26,7 @@
 		</div>
 		<div data-o-grid-colspan="8">
 			<h4>
-				<label>
-					Extract <input id="extract__numberOfEvents-range" type="range" min="100" max="10000" value="100" step="100" oninput="outputUpdate(value)"> <output for="extract__numberOfEvents-range" class="extract__numberOfEvents">100</output> events.
-				</label>
-				<button type="button" class="o-buttons extract__submit">Loading ...</button>
+				<label>{{>range}}</label>
```
Comment (adambraimbridge) [inline views/extract.html]: idea: turn this into a template type thingo

File: views/partials/chart.html [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<div class="chart" data-chart-name="{{name}}" data-o-grid-colspan="{{colspan}}">
+<div class="chart" data-chart-name="{{name}}" data-o-grid-colspan="5">
```
Comment (adambraimbridge) [inline views/partials/chart.html]: {{colspan}} <-- i guess maybe set the colspan in the js instead

File: views/partials/chart.html [hunk 10]
```diff
@@ -16,15 +16,21 @@
 	<div class="chart__printer core-chart"></div>
 
 	<nav class="chart__ui__links">
-		<button class="chart__ui__copy-data o-buttons o-buttons--small">Copy data to clipboard</button>
+		<!--button class="chart__ui__copy-data o-buttons o-buttons--small">Copy data to clipboard</button-->
```
Comment (adambraimbridge) [inline views/partials/chart.html]: good. this has been broken for a while. I say we change it to open the csv data in a new tab for copy and/or file->save

File: views/partials/range.html [hunk 11]
```diff
@@ -0,0 +1,2 @@
+Extract <input class="extract__numberOfEvents-range" type="range" min="100" max="5000" value="1000" step="100"> <output for="extract__numberOfEvents-range" class="extract__numberOfEvents">5000</output> events.
```
Comment (adambraimbridge) [inline views/partials/range.html]: coooool, let's use this partial in https://beacon.ft.com/data/extract/

File: views/recordings.html [hunk 12]
```diff
@@ -0,0 +1,30 @@
+<table class="o-table o-table--compact recordings" data-o-component="o-table">
+	<thead>
+		<th data-sort="string">Location</th>
+		<th data-sort="int">Duration (min)</th>
+		<th data-sort="int">Pages</th>
+		<th data-sort="int">Last Seen</th>
+		<th data-sort="string">Entry Page</th>
+		<th data-sort="string">SpoorId</th>
```
Comment (adambraimbridge) [inline views/recordings.html]: spoorid <> referrer

File: views/recordings.html [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+		{{#if spoorId}}
+		<tr>
+			<td>
+				<a href="{{url}}" target="_blank">
+					<span class="recordings--flag">{{flag}}</span>{{#if city}}{{city}}{{else}}{{country}}{{/if}}
+				</a>
+			</td>
+			<td>{{duration}}</td>
+			<td>{{pages}}</td>
+			<td>TBD</td><!-- {{lastSeen}} -->
```
Comment (adambraimbridge) [inline views/recordings.html]: hide, or format

File: views/partials/chart.html (hunk 9 above)
Comment (lc512k) [inline views/partials/chart.html]: Left it alone for the dashboard view, and forced 5 for the standalone version only

**Comment by lc512k** [comment]
@adambraimbridge Changed everything with a :+1:, put the rest in https://github.com/Financial-Times/next-beacon-v2/issues/277

**Comment by adambraimbridge** [comment]
Great stuff 💯

**Pull request merged**